src/transformers/models/blip_2/modeling_blip_2.py

```diff
@@ -995,6 +1196,9 @@ def __init__(self, config: Blip2QFormerConfig):
         super().__init__(config)
         self.config = config
 
+        if config.use_qformer_text_input:
+            self.embeddings = Blip2TextEmbeddings(config)
```

Comment:
Instead of using this config argument to conditionally call and create this layer, I'd suggest instead call  `self.embeddings` if `input_ids` is not None

```suggestion
        self.embeddings = Blip2TextEmbeddings(config)
```

Reply:
when this layer is created_always_, I got this type of errors, don't know how to fix them.
Some Blip2 models do not use this bert based embeddings, they use _opt_ or _flan-t5_ to create the query_embeds. Maybe I could try to refactor the code to move the `Blip2TextEmbeddings` outside of `Blip2QFormerModel` and pass always query_embeds. what do you think?

```
FAILED tests/models/blip_2/test_modeling_blip_2.py::Blip2ForConditionalGenerationDecoderOnlyTest::test_training_gradient_checkpointing - AssertionError: False is not true : qformer.embeddings.word_embeddings.weight in Blip2ForConditionalGeneration has no gradient!
```

```
FAILED tests/models/blip_2/test_modeling_blip_2.py::Blip2ModelTest::test_training_gradient_checkpointing - AssertionError: False is not true : qformer.embeddings.word_embeddings.weight in Blip2ForConditionalGeneration has no gradient!
```

Reply:
I did a refactor, embeddings were removed from `Blip2QFormerModel`, and place them into `Blip2ForImageTextRetrieval` and `Blip2TextModelWithProjection`, but to do so i needed to add `query_length` param to `Blip2QFormerModel.forward`.